File: rmi/RMIReaderWriter/src/Application/WLocker.java
package Application;

import java.util.concurrent.Semaphore;

public class WLocker implements RWLockable {
	Resource r;
	int readersCounter = 0, writersCounter = 0;
	Semaphore readersCounterMutex = new Semaphore(1), // protects readersCounter
			writersCounterMutex = new Semaphore(1), // protects writersCounter
			closeToWriters = new Semaphore(1),
			closeToReaders = new Semaphore(1);
	
	public WLocker(String name) {
		r = new Resource(name);
	}

	@Override
	public Resource getResource() {
		return r;
	}

	@Override
	public void requestRead() throws InterruptedException {
		closeToReaders.acquire();
		readersCounterMutex.acquire();
		
		if(++readersCounter == 1) 
			closeToWriters.acquire();
		
		readersCounterMutex.release();
		closeToReaders.release();
	}

	@Override
	public void releaseRead() throws InterruptedException {
		readersCounterMutex.acquire();
		
		if(--readersCounter == 0) 
			closeToWriters.release();
		
		readersCounterMutex.release();
	}

	@Override
	public void requestWriting() throws InterruptedException {
		writersCounterMutex.acquire();
		
		if (++writersCounter == 1)
			closeToReaders.acquire(); // impede a entrada de leitores na região crítica
		
		writersCounterMutex.release();
		closeToWriters.acquire(); // impede escritores simultâneos na região crítica
	}

	@Override
	public void releaseWriting() throws InterruptedException {
		closeToWriters.release();
		writersCounterMutex.acquire();
		
		if (--writersCounter == 0)
			closeToReaders.release();
		
		writersCounterMutex.release();
	}

}

File: druby/ReaderWriter/Application/wlocker.rb
require_relative 'resource.rb'

class WLocker
  def initialize(filepath)
    @resource = initialize_resource(filepath)

    @readers_counter = 0
    @writers_counter = 0

    @readers_counter_mutex = Mutex.new # protects @readers_counter
		@writers_counter_mutex = Mutex.new # protects @writers_counter
    @readers_locker_mutex = Mutex.new # Protects readers locker
  	@close_to_writers = Mutex.new
  	@close_to_readers = Mutex.new
  end

  def access_read
    @readers_locker_mutex.synchronize do
      @close_to_readers.synchronize do
        @readers_counter_mutex.synchronize do
          @close_to_writers.lock if((@readers_counter += 1) == 1)
        end
      end
    end

    yield(@resource)

    @readers_counter_mutex.synchronize do
      @close_to_writers.unlock if((@readers_counter -= 1) == 0)
    end
  end

  def access_write
    @writers_counter_mutex.synchronize do
      @close_to_readers.lock if ((@writers_counter += 1) == 1) # impede a entrada de leitores na região crítica
    end

    @close_to_writers.synchronize do
      yield(@resource)
    end # impede escritores simultâneos na região crítica

    @writers_counter_mutex.synchronize do
      @close_to_readers.unlock if ((@writers_counter -= 1) == 0)
    end
  end

  private
  def initialize_resource(filepath="")
    raise NoNameForFileError, 'There is no file with empty name' unless filepath && !filepath.empty?
    Resource.new(filepath)
  end
end
